ext/php7/configuration.c

```diff
@@ -1,306 +1,147 @@
-            return true;
-        } else if (strcmp(subject.ptr, "0") == 0) {
-            return false;
-        }
-    } else if ((subject.len == 4 && strcasecmp(subject.ptr, "true") == 0)) {
-        return true;
-    } else if ((subject.len == 5 && strcasecmp(subject.ptr, "false") == 0)) {
-        return false;
-    }
-    return default_value;
-}
-
-bool ddtrace_config_env_bool(ddtrace_string env_name, bool default_value) {
-    ddtrace_string env_val = ddtrace_string_getenv(env_name.ptr, env_name.len);
-    bool result = default_value;
-    if (env_val.len) {
-        result = ddtrace_config_bool(env_val, default_value);
-    }
-    if (env_val.ptr) {
-        efree(env_val.ptr);
-    }
-    return result;
-}
-
-bool ddtrace_config_distributed_tracing_enabled(void) {
-    ddtrace_string env_name = DDTRACE_STRING_LITERAL("DD_DISTRIBUTED_TRACING");
-    return ddtrace_config_env_bool(env_name, true);
-}
-
-// Get env name for <PREFIX_><INTEGRATION><_SUFFIX> (i.e. <DD_TRACE_><LARAVEL><_ENABLED>)
-size_t ddtrace_config_integration_env_name(char* name, const char* prefix, ddtrace_integration* integration,
-                                           const char* suffix) {
-    ZEND_ASSERT(strlen(prefix) <= DDTRACE_LONGEST_INTEGRATION_ENV_PREFIX_LEN);
-    ZEND_ASSERT(strlen(suffix) <= DDTRACE_LONGEST_INTEGRATION_ENV_SUFFIX_LEN);
-    return (size_t)snprintf(name, DDTRACE_LONGEST_INTEGRATION_ENV_LEN + 1, "%s%s%s", prefix, integration->name_ucase,
-                            suffix);
-}
+// Allow for partially defined struct initialization here
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
 
-// Get env value for <PREFIX_><INTEGRATION><_SUFFIX>
-ddtrace_string _dd_env_integration_value(const char* prefix, ddtrace_integration* integration, const char* suffix) {
-    char name[DDTRACE_LONGEST_INTEGRATION_ENV_LEN + 1];
-    size_t len = ddtrace_config_integration_env_name(name, prefix, integration, suffix);
-    return ddtrace_string_getenv(name, len);
-}
+#define CALIAS_EXPAND(name) {.ptr = name, .len = sizeof(name) - 1},
+#define CALIASES(...) ((zai_string_view[]){APPLY_N(CALIAS_EXPAND, ##__VA_ARGS__)})
+#define CONFIG(type, name, ...) ZAI_CONFIG_ENTRY(DDTRACE_CONFIG_##name, name, type, __VA_ARGS__),
+#define CALIAS(type, name, ...) ZAI_CONFIG_ALIASED_ENTRY(DDTRACE_CONFIG_##name, name, type, __VA_ARGS__),
+static zai_config_entry config_entries[] = {DD_CONFIGURATION};
+#undef CALIAS
+#undef CONFIG
 
-#define DD_INTEGRATION_ENABLED_DEFAULT true
+bool runtime_config_first_init = false;
 
-// note: only call this if ddtrace_config_trace_enabled() returns true
-bool ddtrace_config_integration_enabled(ddtrace_string integration_str) {
-    ddtrace_integration* integration = ddtrace_get_integration_from_string(integration_str);
-    if (integration == NULL) {
-        return DD_INTEGRATION_ENABLED_DEFAULT;
+static void dd_copy_tolower(char *dst, const char *src) {
+    while (*src) {
+        *(dst++) = tolower(*(src++));
```

Comment:
```suggestion
static void dd_copy_tolower(char *restrict dst, const char *restrict src) {
    while (*src) {
        *(dst++) = tolower((unsigned char)*(src++));
```
Restrict is an optimization, but unsigned char cast is needed to avoid undefined behavior. See https://stackoverflow.com/a/21805970/538216 for an explanation.

-----

Note that this will not be an efficient lowercasing routine; gcc and clang seem to be unable to inline `tolower` and issue a function call. Would be better to use something like this instead of `tolower` (untested):

```c
static char dd_ascii_tolower(char in) {
    int adjust = (in >= 'A' && in <= 'Z') ? 'A' - 'a' : 0;
    int i = (unsigned char)in;
    return (char)(i - adjust);
}
```

Of course, I'm not sure it would matter. This will get called a few times, but only during MINIT or first-time RINIT, I think?

Reply:
(not undefined as the inputs are all basic ASCII)

I guess using this `- ('A' - 'a')` will result in a more efficient asm at least.

Reply:
https://godbolt.org/z/qdY8a1fos - not sure what I like more :-D (the gcc output for the lower one is inefficient though)

Reply:
BTW, passing in length information allows it to vectorize with the one I provided (and I would guess yours as well).